# Lecture 3
- HW1 due tomorrow
- HW2 posted yesterday and will be due the upcoming sunday

## Dictionary
- Dictionary allows us to index using arbitrary datatype (i.e. a string) instead of only integers.
- Dictionary contains **key-value pairs** in the format of **key:val**

One way to define dictionary:

In [2]:
dict1 = {"fruit":"apple", "animal":"cat", "ai":"chatgpt"}

Access values stored in each keys:

In [3]:
dict1["fruit"]

'apple'

Another way to define a dictionary

In [6]:
dict2 = {}
dict2 ["fruit"] = "cherry"
dict2 ["ai"] = "gemini"

In [7]:
print(dict1, dict2)

{'fruit': 'apple', 'animal': 'cat', 'ai': 'chatgpt'} {'fruit': 'cherry', 'ai': 'gemini'}


you can also change the existing values in dictionaries using similar manner

In [9]:
dict2["ai"] = "claude"

In [10]:
print(dict2)

{'fruit': 'cherry', 'ai': 'claude'}


## Loop through a dictionary
- loop through the keys

In [11]:
dict1_keys = dict1.keys()
print(dict1_keys)

dict_keys(['fruit', 'animal', 'ai'])


In [13]:
for key in dict1_keys:
    val = dict1[key]
    print(val)

apple
cat
chatgpt


- you can also directly loop though the entire dictionary using **.items()**:

In [14]:
for key, val in dict1.items():
    print(f"key is {key} and value is {val}")

key is fruit and value is apple
key is animal and value is cat
key is ai and value is chatgpt


## Basic Operations and Boolean
in python and other prog. languages, a single = sign does not mean "equal to", it means "assigned to", so what is the sign that means "equal to"?

In [15]:
5==4

False

In [16]:
5==5

True

In [17]:
[1,2,3] == [3,2,1]

False

In [18]:
[1,2,3] == [1,2,3]

True

Another important aspect is inequality:
- _> greater than_
- _< less than_
- _>= greater than and equal to_
- _<= less than and equal to_
  

In [19]:
print(5>3, 5>5, 5>=5, 5<=5)

True False True True


Another "less intuitive" logic is "not equal to", so this means I want to know if the object on the left is "not the same" as the one on the right. This is done with "!=":

In [20]:
print(5!=3, 5!=5)

True False


These true/false statements are very useful in a new structure called **if loop**

In [23]:
Is_my_hair_too_long = True
if Is_my_hair_too_long:
    print("Get a haircut!")
else:
    print("get a haircut later")

Get a haircut!


You can also compose more complex logic with **and** and **or** statement:
- **and** only returns True if both boolean on the left and right is True
- **or** return True as long as one boolean on the left or right is true

In [24]:
print(1==1 and 2==1)

False


In [25]:
print(1==1 and 2==2)

True


In [26]:
print(1==1 or 2==1)

True


In [27]:
print(1==1 or 2==2)

True


In [28]:
print(1==2 or 2==1)

False


Combining if loop and for loop for more complex operation

In [29]:
for item in range(1,10):
    print(item)

1
2
3
4
5
6
7
8
9


In [30]:
for item in range(1,10):
    if item == 7:
        print("yay")
    else:
        print("nay")

nay
nay
nay
nay
nay
nay
yay
nay
nay


When we store True and False in python, they are actually stored as 0s and 1s.

In [31]:
print(int(True), int(False))

1 0


In [32]:
print(True+True, True+False)

2 1


This is very useful in counting entries in a for loop:
- suppose I want to write a for loop that counts how many cats are in an array:

In [33]:
animals = ['cat', 'dog', 'dog', 'cat', 'dog', 'bird', 'dog', 'cat']
cat_count = 0
for item in animals:
    cat_count = cat_count + (item == "cat")
print(cat_count)

3


## Basic Math Operation
- power

In [34]:
print(2*4, 2**4) #the first is "two times four", the second is "two to the power of 4"

8 16


In [35]:
print(2**16)

65536


In [36]:
print(2**(1/2))

1.4142135623730951


Using for loop to add the values in a list

In [37]:
costs = [1.2, 3.4, 4.5, 4., 6.]
total = 0
for cost in costs:
    total += cost # total += cost is the same as total = total + cost
print(total)

19.1


The += sign can be generalized to -=,*=, /=:
- x -= y is x = x-y
- x *= y is x = x*y

In [38]:
cash = 100
for cost in costs:
    cash -= cost
print(cash)

80.89999999999999


In [39]:
factorial = 1
for i in range(5): # looping through 0, 1, 2, 3, 4
    factorial *= i
print(factorial)

0


In [41]:
factorial = 1
for i in range(5): # looping through 0, 1, 2, 3, 4
    factorial *= (i+1) # i+1 is actually 1,2,3,4,5
print(f"5! is equal to {factorial}")

5! is equal to 120


You need to be very careful about 0s in your *= /=

In [42]:
inverse_factorial = 1
for i in range(5):
    inverse_factorial /= i
print(inverse_factorial)

ZeroDivisionError: division by zero

In [43]:
inverse_factorial = 1
for i in range(5):
    inverse_factorial /= (i+1)
print(inverse_factorial, 1/inverse_factorial)

0.008333333333333333 120.0


## Enumerate
remember we discussed about two ways to loop through a list:
- for item in list1:
- for index in range(len(list1))
if you loop through a list with range(len(list1)) command, then you have the extra benefit of knowing the index of each step.

But, if you want to loop through a list using the first way (i.e. for item in list1:), what's the best way to know the index?

In [47]:
short_list = ["chatgpt", "claude", "gemini", "deepseek"]

In [49]:
for i, llm in enumerate(short_list):
    print(f"The large language model at index {i} is {llm}") 

The large language model at index 0 is chatgpt
The large language model at index 1 is claude
The large language model at index 2 is gemini
The large language model at index 3 is deepseek


## While loop
while loop will keep running until its condition is broken.

In [50]:
i = 10
while i > 5:
    print(i)
    i = i - 1 # this line is very IMPORTANT

10
9
8
7
6


You can write a while loop using more complex logic

In [51]:
count = 0
i = 0
while (i**2 < 30 and count <10):
    count += 1
    i+=1
    print(f"i is {i} and count is {count}")

i is 1 and count is 1
i is 2 and count is 2
i is 3 and count is 3
i is 4 and count is 4
i is 5 and count is 5
i is 6 and count is 6


You can also "manually" stop the while loop using **break** command:

In [52]:
while True: #You should never write this line of code (unless you know what you are doing)
    print("this while loop will never stop by itself! too bad")
    break

this while loop will never stop by itself! too bad


## Functions

How to define a function

In [53]:
def say_hi(name):
    print("Hi " + name)

In [54]:
say_hi("Prof. Li")

Hi Prof. Li


In [55]:
say_hi(5)

TypeError: can only concatenate str (not "int") to str

it's usually a good habit to
- force the input argument (i.e. name) to be certain format
- set up a warning

In [56]:
def say_hi2(name):
    print("Hi " + str(name))
def say_hi3(name):
    if type(name) == str:
        print("Hi " + name)
    else:
        print("Error: input name is not a string")

In [58]:
say_hi2(5)
say_hi3(5)

Hi 5
Error: input name is not a string


In [60]:
def factorial(n):
    out = 1
    if n > 0: # we need to know if the input is a positive integer for factorial to be well-defined
        for i in range(int(n)):
            out *= (i+1)
    return out

Return statement is useful because it immediately terminate the function

In [61]:
print(factorial(5), factorial(7), factorial(6.6))

120 5040 720


720 is actually 6!, so basically the .6 part of 6.6 is ignored. This make sense because in my factorial function I convert any floating point number into integers

To calculate the factorial of fractional number, we need to use gamma function.

Gamma[x] = (x-1)Gamma[x-1]

and 

Gamma[1] = 1

So we know that Gamma[n] = (n-1)!

In [64]:
def Gamma(n):
    if n <= 1:
        return 1
    else:
        return (n-1)*Gamma(n-1) #This is recursion

**Recursion** means a function calls itself

In [66]:
print(Gamma(5+1), Gamma(3.2))

120 0.5280000000000006


## Sorting algorithm

In [67]:
def dumb_sort(alist):

    b = [] # I'm making a place to put my numbers in order, this is the out list

    c = alist[:] # I'm making a copy of my original list

    while len(c)>0: #I'm going to keep removing numbers from my c list and put them into the b list until c is empty

        largest = c[0]

        for item in c: # run over all elements in the c list
            if (item > largest): # if the current element (item) is larger than the stored largest value
                largest = item # replace largest with the current element
        b.append(largest) # adding the largest values figured out from the c list
        c.pop(c.index(largest)) # removing the largest values from the c list, this is important because our while loop is based on c list size
    return b

In [68]:
unordered = [3,5,1,5,2,4,5,2,1]

In [69]:
dumb_sort(unordered)

[5, 5, 5, 4, 3, 2, 2, 1, 1]

In [73]:
import time
tstart = time.time()
dumb_sort(unordered*10000)
print(time.time() - tstart)

KeyboardInterrupt: 